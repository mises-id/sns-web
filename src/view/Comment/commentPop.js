/*
 * @Author: lmk
 * @Date: 2022-01-10 16:23:16
 * @LastEditTime: 2022-05-17 10:20:50
 * @LastEditors: lmk
 * @Description:
 */
import React, {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useState,
} from "react";
import "./index.scss";
import replies_close from "@/images/replies_close.png";
import replies_refresh from "@/images/replies_refresh.png";
import { Popup } from "zarm";
import {
  formatTimeStr,
  isMe,
  objToUrl,
  useBind,
  useList,
  useLoginModal,
  username,
  useRouteState,
} from "@/utils";
import liked from "@/images/liked.png";
import like from "@/images/like.png";
import { getComment } from "@/api/status";
import PullList from "@/components/PullList";
import deleteComment from "@/images/deleteComment.png";
import ReplyInput from "@/components/ReplyInput";
import { useRef } from "react";
import { useSelector } from "react-redux";
import { createComment } from "@/api/comment";
import { useHistory } from "react-router-dom";
import Avatar from '@/components/NFTAvatar'
const CommentsPop = (
  {
    setvisible,
    visible,
    comment,
    likePress,
    createdUserId,
    deleteCommentData,
    submit,
    setParentSelectItem,
    inputContent,
  },
  ref
) => {
  const state = useRouteState(); // get router a query
  const [lastId, setlastId] = useState("");
  const status_id = state.id;
  const [data, setdata] = useState(comment);// comment data
  const [isSetTop, setisSetTop] = useState(false); // Has the first level evaluation been set
  const user = useSelector((state) => state.user) || {};
  const commentContent = useBind("");
  const input = useRef();
  const [selectItem, setselectItem] = useState(data); // select a comment item
  const { avatar = {} } = data.user || {};
  const [setrefreshFlag, setrefresh] = useState(false); // refresh status
  const loginModal = useLoginModal();
  const history = useHistory()
  const idForm = {}
  if(comment.nft_asset_id){
    idForm.nft_asset_id = comment.nft_asset_id
  }
  if(comment.status_id){
    idForm.status_id = status_id || comment.status_id
  }
  let [fetchData, last_id, dataSource, setdataSource] = useList(getComment, {
    topic_id: comment.id,
    limit: 20,
    last_id: lastId,
    ...idForm,
  });
  useEffect(() => {
    setlastId(last_id);
  }, [last_id]);
  const [setTopCommentId, setsetTopCommentId] = useState('');
  useEffect(() => {
    setselectItem(comment);
    setdata(comment)
  }, [comment]);
  useImperativeHandle(ref, () => {
    return {
      // update datalist
      setData(res) {
        if(res){
          console.log(res);
          dataSource.unshift(res);
          setdataSource([...dataSource]);
        }
      },
      removeItem(valId){
        const valIndex = dataSource.findIndex(val=>val.id===valId)
        dataSource.splice(valIndex,1)
        dataSource.length===0 ? fetchData('refresh') : setdataSource([...dataSource])
      },
      setTopCommentId(id){
        setsetTopCommentId(id)
      }
    };
  });
  // Prepare datalist
  useEffect(()=>{
    if(dataSource.length>0&&!isSetTop){
      const first = dataSource[0];
      const lastList = dataSource.slice(1);
      const findIndex = lastList.findIndex(val=>val.id===setTopCommentId);
      if(findIndex>-1){
        lastList.splice(findIndex,1);
        setisSetTop(true)
        // eslint-disable-next-line
        dataSource = [first,...lastList]
        setdataSource([...dataSource])
      }
    }
  },[dataSource.length])
  // Monitor pop-up status
  useEffect(() => {
    return () => {
      if (visible) {
        setdataSource([]);
        setlastId("");
      }
    };
    // eslint-disable-next-line
  }, [visible]);
  // Select the content to reply to
  const selectReplyItem = (val) => {
    commentContent.onChange("");
    val.username = username(val.user);
    setselectItem(val);
    setParentSelectItem&&setParentSelectItem(val);
    input.current && input.current.focus();
  };
  // click item
  const replyItem = (val) => {
    if (!user.token) {
      loginModal(() => {
        selectReplyItem(val);
      });
      return false;
    }
    selectReplyItem(val);
  };
  // route to user detail
  const userDetail = (e,{user:item})=>{
    e.stopPropagation();
    const isMe = user.loginForm.uid === item.uid;
    if (!isMe) {
      setvisible(false)
      setTimeout(() => {
        const avatar = item.avatar ? item.avatar.medium : "";
        history.push({
          pathname: "/userDetail",
          search: objToUrl({ uid: item.uid, username: item.username, avatar,is_followed: item.is_followed,misesid:item.misesid }),
        });
      }, 10);
    }
  }
  // render data
  const renderView = (val, index) => {
    const valAvatar = val.user.avatar || {}
    return (
      <div
        className="m-flex m-col-top m-padding-top13 m-padding-left15 child-comment"
        key={index}
        onClick={() => replyItem(val)}
      >

        <Avatar size="30px" avatarItem={valAvatar} onClick={e=>userDetail(e,val)}  key={val.id} />
        {/* <Image size={30} source={valAvatar.medium}></Image> */}
        <div className="m-margin-left11 m-line-bottom m-flex-1">
          <span className="commentNickname">{username(val.user)}</span>
          <div className="m-font15 m-colors-555 m-margin-top8 right-content  m-padding-bottom13">
            <div className="comment-content">
              {val.opponent && (
                <>
                  Reply to{" "}
                  <span className="at-name">@{username(val.opponent)}</span>:
                </>
              )}
              {val.content}
            </div>
            <div className="m-flex m-row-between m-margin-top8">
              <div className="m-flex m-row-center">
                <span className="m-colors-666 m-font12 m-margin-right8">
                  {formatTimeStr(val.created_at)}
                </span>
                {isMe(val.user, createdUserId) && (
                  <img
                    src={deleteComment}
                    alt=""
                    width={12}
                    onClick={(e) => deleteCommentData(e, val)}
                  />
                )}
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  };
  // parent comment
  const otherView = () => {
    return (
      <div
        className="m-flex m-col-top m-padding-top13 m-bg-fff m-padding-left15"
        onClick={() => replyItem(data)}
      >
        <Avatar size="30px" avatarItem={avatar} onClick={e=>userDetail(e,data)}/>
        {/* <Image size={30} source={avatar && avatar.medium} onClick={e=>userDetail(e,data)}></Image> */}
        <div className="m-margin-left11 m-line-bottom m-flex-1">
          <span className="commentNickname">{username(data.user)}</span>
          <div className="m-font15 m-colors-555 m-margin-top8   m-padding-bottom13">
            <p className="comment-content">{data.content}</p>
            <div className="m-flex m-row-between m-margin-top8">
              <div className="m-flex m-row-center">
                <span className="m-colors-666 m-font12 m-margin-right8">
                  {formatTimeStr(data.created_at)}
                </span>
                {isMe(data.user, createdUserId) && (
                  <img
                    src={deleteComment}
                    alt=""
                    width={12}
                    onClick={(e) => deleteCommentData(e, data)}
                  />
                )}
              </div>
              <div className="right-icon m-flex">
                <div
                  className="m-flex like-box"
                  onClick={(e) => likePress(e, data)}
                >
                  <img
                    src={data.is_liked ? liked : like}
                    width={13}
                    alt="like"
                  ></img>
                  <span
                    className={`m-font11 m-margin-left8 ${
                      data.is_liked ? "m-colors-FF3D62" : "m-colors-333"
                    }`}
                  >
                    {data.likes_count || 0}
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  };
  // refresh data list
  const refresh = () => {
    if (!setrefreshFlag) {
      fetchData("refresh");
      setrefresh(true);
      setTimeout(() => setrefresh(false), 2500);
    }
  };
  // close this pop
  const closePop = () => {
    setvisible(false);
    commentContent.onChange("");
    setselectItem("");
    setisSetTop(false)
  };
  // click submit button
  const submitItem = (e) => {
    inputContent&&inputContent.onChange(commentContent.value);
    submit&&submit(e);
    createComment({
      content: commentContent.value,
      ...idForm,
      parent_id: selectItem.id || "",
    }).then((res) => {
      commentContent.onChange("");
      setselectItem(data);
      dataSource.unshift(res);
      setdataSource([...dataSource]);
      data.comments_count+=1
      setdata({...data})
    });
  };
  return (
    <Popup
      visible={visible}
      onMaskClick={closePop}
      direction="bottom"
      mountContainer={() => document.body}
    >
      <div>
        <div className="title-box m-flex">
          <div
            className={`pop-icon refresh ${
              setrefreshFlag ? "refresh-active" : ""
            }`}
            onClick={refresh}
          >
            <img src={replies_refresh} alt="" width={18} />
          </div>
          <div className="m-flex-1 pop-title">
            {data.comments_count} Replies
          </div>
          <div className="pop-icon" onClick={closePop}>
            <img src={replies_close} alt="" width={18} />
          </div>
        </div>
        <div className="pop-content-scroll">
          <PullList
            renderView={renderView}
            data={dataSource}
            otherView={otherView}
            load={fetchData}
          ></PullList>
        </div>
        <ReplyInput
          submit={submitItem}
          content={commentContent}
          ref={input}
          defaultItem={data}
          setselectItem={setselectItem}
          placeholder={
            selectItem && selectItem.username ? `@${selectItem.username}` : ""
          }
        ></ReplyInput>
      </div>
    </Popup>
  );
};
export default forwardRef(CommentsPop);
