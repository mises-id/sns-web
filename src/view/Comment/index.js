/*
 * @Author: lmk
 * @Date: 2021-07-15 16:07:01
 * @LastEditTime: 2022-01-12 18:23:07
 * @LastEditors: lmk
 * @Description: comment
 */

import "./index.scss";
import React, { useEffect, useRef, useState } from "react";
import { useTranslation } from "react-i18next";
import { Toast } from "zarm";
import PullList from "@/components/PullList";
import Navbar from "@/components/NavBar";
import { formatTime, useBind, useList, username, useRouteState } from "@/utils";
import { getComment } from "@/api/status";
import { createComment, likeComment, unlikeComment } from "@/api/comment";
import { useSelector } from "react-redux";
import liked from '@/images/liked.png'
import like from '@/images/like.png'
import commentIcon from '@/images/comment.png'
import ReplyInput from "@/components/ReplyInput";
import Image from "@/components/Image";
import CommentsPop from "./commentPop";

const Comment = ({ history }) => {
  const { t } = useTranslation();
  const state = useRouteState();
  const [selectItem, setselectItem] = useState({})
  const input = useRef();
  const likePress = (e,val)=>{
    e.stopPropagation()
    const fn = val.is_liked ? unlikeComment : likeComment;
    fn(val.id).then(res=>{
      val.is_liked = !val.is_liked;
      const findIndex = dataSource.findIndex(item=>item.id===val.id);
      val.likes_count = val.is_liked ? ++val.likes_count : --val.likes_count
      if(findIndex>-1){
        dataSource[findIndex] = val;
        console.log(dataSource)
        setdataSource([...dataSource])
      }
    }).catch(res=>{
      
    })
  }
  const replyItem = (val,e)=>{
    if(e){
      e.stopPropagation();
    }
    commentContent.onChange('')
    val.username = username(val.user)
    setselectItem(val)
    input.current.focus()
  }
  const renderView = (val = {}, index) => {
    const {user = {},comments=[]} = val;
    const {avatar={}} = user;
    return (
      <div
        key={index}
        className="m-flex m-col-top m-padding-top13 m-bg-fff m-padding-left15"
        onClick={()=>replyItem(val)}
      >
        <Image
          size={30}
          source={avatar&&avatar.medium}
        ></Image>
        <div className="m-margin-left11 m-line-bottom m-flex-1">
          <span className="commentNickname">
            {username(user)}
          </span>
          <div className="m-font15 m-colors-555 m-margin-top8 right-content  m-padding-bottom13">
            <p className="comment-content">{val.content}</p>
            <div className="m-flex m-row-between m-margin-top8">
              <span className="m-colors-666 m-font12">{formatTime((val.created_at))}</span>
              <div className="right-icon m-flex">
                <div className="m-flex like-box"  onClick={(e)=>likePress(e,val)}>
                  <img src={val.is_liked ? liked : like}  className='icon' alt="like"></img>
                  <span className={`m-font14 m-margin-left8 ${val.is_liked ? 'm-colors-FF3D62' : 'm-colors-333'}`}>{val.likes_count || 0}</span>
                </div>
                <div className="m-flex icon-box">
                  <img src={commentIcon}  className='icon' alt="comment"></img>
                  <span className={`m-font14 m-colors-333 m-margin-left8`}>{val.comments_count || 0}</span>
                </div>
              </div>
            </div>
          </div>
          <div className="all-comment">
            {comments.map((item,i)=>{
              const {user={},content,created_at} = item;
              const {avatar={}} = user;
              return <div
                  key={i}
                  className="m-flex m-col-top m-bg-fff m-padding-bottom15 right-content"
                  onClick={(e)=>replyItem(item,e)}
                >
                  <Image
                    size={20}
                    source={avatar&&avatar.medium}
                  ></Image>
                  <div className="m-margin-left11 m-flex-1">
                    <div className="m-padding-bottom10">
                      <span className="commentNickname1">{username(user)}{item.opponent&&<span className="at-name">@{username(item.opponent)}</span>}:</span>
                      <p className="comment-content1">{content}</p>
                    </div>
                    <span className="m-colors-666 m-font12">{formatTime((created_at))}</span>
                  </div>
                </div>
            })}
            {val.comments_count>3&&<p className="more-comment-list" onClick={(e)=>showMoreComment(e,val)}>View all {val.comments_count - comments.length} comments</p>}
          </div>
        </div>
      </div>
    );
  };
  //getData
  const [lastId, setlastId] = useState("");
  const [fetchData, last_id, dataSource, setdataSource] = useList(getComment, {
    status_id: state.id,
    limit: 20,
    last_id: lastId,
  });
  const user = useSelector((state) => state.user) || {};
  const [loading, setloading] = useState(false);
  useEffect(() => {
    setlastId(last_id);
  }, [last_id]);
  const commentContent = useBind("");
  const commentsPop = useRef('')
  const submit = (e) => {
    e.preventDefault();
    if (!user.token) {
      Toast.show(t("notLogin"));
      return false;
    }
    if (loading || !commentContent.value) {
      return false;
    }
    setloading(true);
    createComment({
      content: commentContent.value,
      status_id: state.id,
      parent_id:selectItem.id || '',
    })
      .then((res) => {
        if(selectItem.content){ // if has select item
          if(selectItem.comments_count<3){
            selectItem.comments.push(res);
            const findIndex = dataSource.findIndex(val=>val.id===selectItem.id)
            if(findIndex>-1){
              dataSource[findIndex] = selectItem;
            }
          }
          // if popup is show flag
          if(visible){
            commentsPop.current.setData(res)
          }
          selectItem.comments_count+=1;
          setselectItem('')
        }else{
          dataSource.unshift(res);
        }
        setdataSource([...dataSource]);
        commentContent.onChange("");
      })
      .finally(() => {
        setloading(false);
      });
  };
  const [visible, setvisible] = useState(false)
  const [commentPop, setcommentPop] = useState({})
  const showMoreComment = (e,val)=>{
    e.stopPropagation()
    setvisible(true)
    setcommentPop(val)
    val.username = username(val.user)
    setselectItem(val)
  }
  return (
    <div className="m-flex m-flex-col page">
      <Navbar
        title={
          t("commentPageTitle") +
          (dataSource.length ? `(${dataSource.length})` : "")
        }
      />
      <div className="m-flex-1 commentBox">
        <PullList
          renderView={renderView}
          data={dataSource}
          load={fetchData}
        ></PullList>
      </div>
      <CommentsPop visible={visible} ref={commentsPop} setvisible={setvisible} comment={commentPop} replyItem={replyItem} likePress={likePress}></CommentsPop>
      <ReplyInput ref={input} submit={submit} content={commentContent} placeholder={selectItem&&selectItem.username ? `@${selectItem.username}` : ''}></ReplyInput>
    </div>
  );
};
export default Comment;
