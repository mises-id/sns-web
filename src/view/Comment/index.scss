.footer{
  padding: 15px 20px;
}
.commentBox{
  .m-layout{
    height: calc(100vh - 45px - 60px);
  }
}
.commentNickname{
  @include title(15);
  color:#111
}
.commentNickname1{
  @include title(14);
  color:#111
}
.content{
  margin-right: 24px;
  line-height: 19px;
}
.comment{
  margin-left: 10px;
  height: 40px;
  border-radius: 500px;
  background-color: $themeBg;
}
.icon{
  width: 13px;
  height: 13px;
}
.like-box{
  margin-right: 24px;
}
.right-content{
  padding-right: 30px;
}
.comment-content{
  color: #333;
  font-size: 15px;
}
.comment-content1{
  color: #333;
  font-size: 14px;
}
.more-comment-list{
  margin-bottom: 15px;
  margin-left: 31px;
  color: #5D61FF;
  font-size: 12px;
  line-height: 1;
}
@keyframes ratate {
  form{
    transform: rotate(0);
  }
  to{
    transform: rotate(720deg);
  }
}
.title-box{
  background: white;
  height: 50px;
  border-radius: 15px 15px 0px 0px;
  .pop-icon{
    width: 18px;
    height: 18px;
    margin: 17px;
    img{
      display: block;
      margin: 0 auto;
    }
  }
  .refresh{
    animation-fill-mode:forwards;
  }
  .refresh-active{
    animation: ratate 2s;
  }
  .pop-title{
    text-align: center;
    color: #333333;
    font-size: 17px;
    font-weight: bold;
  }
}
.pop-content-scroll{
  height: 80vh;
  overflow-y:auto ;
  padding-bottom: 60px;
  background-color: white;
  .m-layout{
    height: calc(80vh - 60px);
  }
}
.at-name{
  color: #475884;
  font-weight: 400;
}
.child-comment{
  background: #F8F8F8;
}