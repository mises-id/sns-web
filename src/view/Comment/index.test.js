/*
 * @Author: lmk
 * @Date: 2021-07-15 16:07:05
 * @LastEditTime: 2021-07-15 16:07:36
 * @LastEditors: lmk
 * @Description: Comment test
 */

import React from 'react';
import {shallow } from 'enzyme';
import Comment from './';
describe('Comment', () => {
  it('should render correctly', () => {
    const component = shallow(<Comment />)
    expect(component).toMatchSnapshot();
  });
});