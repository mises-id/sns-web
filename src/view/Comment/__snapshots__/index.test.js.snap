// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment should render correctly 1`] = `
<div
  className="m-flex m-flex-col page"
>
  <Navbar
    title="commentPageTitle"
  />
  <div
    className="m-flex-1 commentBox"
  >
    <PullList
      data={Array []}
      load={[Function]}
      renderView={[Function]}
    />
  </div>
  <ForwardRef(CommentsPop)
    comment={Object {}}
    deleteCommentData={[Function]}
    likePress={[Function]}
    replyItem={[Function]}
    setvisible={[Function]}
    visible={false}
  />
  <ForwardRef(ReplyInput)
    content={
      Object {
        "onChange": [Function],
        "value": "",
      }
    }
    placeholder=""
    submit={[Function]}
  />
</div>
`;
